File: src/features/skills/api/skills.ts
import { invoke, isTauri } from "@tauri-apps/api/core";
import type {
  SkillDetail,
  SkillImportKind,
  SkillImportResult,
  SkillListResult,
  SkillSummary,
} from "../../../types/skill";

export function listSkills(): Promise<SkillListResult> {
  if (!isTauri()) return Promise.resolve({ skills: [], warnings: [] });
  return invoke<SkillListResult>("skills_list");
}

export function getSkillDetail(skillId: string): Promise<SkillDetail> {
  if (!isTauri()) return Promise.reject(new Error("技能详情需要在 Tauri 应用中读取。"));
  return invoke<SkillDetail>("skills_get_detail", { skillId });
}

export function importSkill(
  path: string,
  kind: SkillImportKind,
  replaceExisting = false,
): Promise<SkillImportResult> {
  if (!isTauri()) return Promise.reject(new Error("技能安装需要在 Tauri 应用中执行。"));
  return invoke<SkillImportResult>("skills_import", {
    request: { path, kind, replaceExisting },
  });
}

export function setSkillEnabled(skillId: string, enabled: boolean): Promise<SkillSummary> {
  if (!isTauri()) return Promise.reject(new Error("技能状态需要在 Tauri 应用中保存。"));
  return invoke<SkillSummary>("skills_set_enabled", { skillId, enabled });
}

export function uninstallSkill(skillId: string): Promise<void> {
  if (!isTauri()) return Promise.reject(new Error("技能删除需要在 Tauri 应用中执行。"));
  return invoke<void>("skills_uninstall", { skillId });
}

export function restoreBuiltinSkill(skillId: string): Promise<SkillSummary> {
  if (!isTauri()) return Promise.reject(new Error("技能恢复需要在 Tauri 应用中执行。"));
  return invoke<SkillSummary>("skills_restore_builtin", { skillId });
}

File: src-tauri/src/commands/skills.rs
//! Skill 管理的 Tauri 命令边界。

use tauri::{async_runtime, State};

use crate::{
    skills::types::{
        SkillDetail, SkillImportRequest, SkillImportResult, SkillListResult, SkillSummary,
    },
    state::AppState,
};

#[tauri::command]
pub async fn skills_list(state: State<'_, AppState>) -> Result<SkillListResult, String> {
    let _guard = state.skill_operations.lock().await;
    let repository = state.skill_repository.clone();
    async_runtime::spawn_blocking(move || repository.list())
        .await
        .map_err(|error| format!("技能列表后台任务失败：{error}"))?
}

#[tauri::command]
pub async fn skills_get_detail(
    state: State<'_, AppState>,
    skill_id: String,
) -> Result<SkillDetail, String> {
    let repository = state.skill_repository.clone();
    async_runtime::spawn_blocking(move || repository.get_detail(skill_id.trim()))
        .await
        .map_err(|error| format!("技能详情后台任务失败：{error}"))?
}

#[tauri::command]
pub async fn skills_import(
    state: State<'_, AppState>,
    request: SkillImportRequest,
) -> Result<SkillImportResult, String> {
    let _guard = state.skill_operations.lock().await;
    let repository = state.skill_repository.clone();
    async_runtime::spawn_blocking(move || repository.import(request))
        .await
        .map_err(|error| format!("技能安装后台任务失败：{error}"))?
}

#[tauri::command]
pub async fn skills_set_enabled(
    state: State<'_, AppState>,
    skill_id: String,
    enabled: bool,
) -> Result<SkillSummary, String> {
    let _guard = state.skill_operations.lock().await;
    let repository = state.skill_repository.clone();
    async_runtime::spawn_blocking(move || repository.set_enabled(skill_id.trim(), enabled))
        .await
        .map_err(|error| format!("技能状态后台任务失败：{error}"))?
}

#[tauri::command]
pub async fn skills_uninstall(state: State<'_, AppState>, skill_id: String) -> Result<(), String> {
    let _guard = state.skill_operations.lock().await;
    let repository = state.skill_repository.clone();
    async_runtime::spawn_blocking(move || repository.uninstall(skill_id.trim()))
        .await
        .map_err(|error| format!("技能删除后台任务失败：{error}"))?
}

#[tauri::command]
pub async fn skills_restore_builtin(
    state: State<'_, AppState>,
    skill_id: String,
) -> Result<SkillSummary, String> {
    let _guard = state.skill_operations.lock().await;
    let repository = state.skill_repository.clone();
    async_runtime::spawn_blocking(move || repository.restore_builtin(skill_id.trim()))
        .await
        .map_err(|error| format!("恢复内置技能后台任务失败：{error}"))?
}
